src/type_conversion/jsimport.c

```diff
@@ -1,77 +1,416 @@
+  QUIT_IF_NULL(ModuleSpec);
+  // The ModuleSpec init function takes two arguments, name and loader.
+  // PyModule_FromDefAndSpec uses name for the name of the generated module
+  // It never uses loader argument and doesn't store the ModuleSpec, so it's fine to pass None
+  // for this.
+  spec = PyObject_CallFunctionObjArgs(ModuleSpec, name, Py_None, NULL);
+  QUIT_IF_NULL(spec);
+  module = PyModule_FromDefAndSpec(&JsModule, spec);
+  QUIT_IF_NULL(module);
+  QUIT_IF_NZ(PyModule_ExecDef(module, &JsModule));
+
+  _JsImport_setHiwireObject(module, package_id);
+
+  __dir__ = PyObject_CallFunctionObjArgs(JsImportDirObject, module, NULL);
+  QUIT_IF_NULL(__dir__);
+  module_dict = PyModule_GetDict(module);
+  // make cleanup code more consistent by increfing module_dict.
+  Py_INCREF(module_dict);
+  // "PyDict_SetItem DOES NOT steal a reference to the object"
+  // So that means it increfs the value automatically.
+  QUIT_IF_NZ(_PyDict_SetItemId(module_dict, &PyId___dir__, __dir__));
+
+  jsproxy = _JsImport_getJsProxy(module);
+  // make cleanup code more consistent by increfing jsproxy.
+  Py_INCREF(jsproxy);
+  QUIT_IF_NZ(_PyDict_SetItemId(module_dict, &PyId_jsproxy, jsproxy));
+  QUIT_IF_NZ(PyDict_SetItem(sys_modules, name, module));
+
+  success = true;
+finally:
+  Py_CLEAR(name);
+  Py_CLEAR(sys_modules);
+  Py_CLEAR(importlib_machinery);
+  Py_CLEAR(ModuleSpec);
+  Py_CLEAR(spec);
+  Py_CLEAR(module);
+  Py_CLEAR(__dir__);
+  Py_CLEAR(module_dict);
+  Py_CLEAR(jsproxy);
+  return success ? 0 : -1;
+}
+
+int
+JsImport_dismount(char* name_utf8){
+  bool success = false;
+  PyObject* name = PyUnicode_FromString(name_utf8);
+  PyObject* sys_modules = PyImport_GetModuleDict();
+  if(sys_modules == NULL){
+    goto finally;
+  }
+  PyObject* module = PyDict_GetItemWithError(sys_modules, name);
+  if(module == NULL){
+    PyErr_Format(PyExc_KeyError,
+      "Cannot dismount module '%s': no such module exists.", name_utf8
+    );
+    goto finally;
+  }
+  if(!JsImport_Check(module)){
+    PyErr_Format(PyExc_KeyError,
+      "Cannot dismount module '%s': it was not mounted with 'pyodide.mountPackage',"
+      "rather it is an actual Python module.", name_utf8
+    );
+    goto finally;
+  }
+  if(PyDict_DelItem(sys_modules, name)){
+    goto finally;
   }
 
-  return 0;
+  success = true;
+finally:
+  Py_CLEAR(name);
+  return success ? 0 : -1;
+}
+
+
+int
+JsImport_init()
+{
+  if(
+    EM_ASM_INT({
+      try {
+        Module.mountPackage = function(name, obj){
+          let obj_id = Module.hiwire.new_value(obj);
+          // TODO: Do we need to free name or does the module take ownership of it?
+          let name_utf8 = stringToNewUTF8(name);
+          if(_JsImport_mount(name_utf8, obj_id)){
+            _pythonexc2js();
+          }
+          _free(name_utf8);
+        };
+
+        Module.dismountPackage = function(name){
+          let name_utf8 = stringToNewUTF8(name);
```

Comment:
Would you want to remove the object from hiwire as well?

Reply:
[The hiwire object gets removed automatically when the `JsProxy` is freed.](https://github.com/iodide-project/pyodide/blob/2d33d5ead916146d62c1f30a479db685028a976a/src/type_conversion/jsproxy.c#L33)
However, `JsProxy_cNew` does not steal the reference passed into it, so in `Module.mountPackage` I am leaking `obj_id`. Thanks for checking.

Reply:
These refcounts are really hard work to keep track of.